### PR TITLE
tests: Add fuzzing harness for strprintf(…)

### PR DESCRIPTION
**Pull request by practicalswift**
Add fuzzing harness for `strprintf(…)`.

Update `FuzzedDataProvider.h`.

Avoid hitting some issues in tinyformat (reported upstreams in https://github.com/c42f/tinyformat/issues/70).

---

Found issues in tinyformat:

**Issue 1.** The following causes a signed integer overflow followed by an allocation of 9 GB of RAM (or an OOM in memory constrained environments):

```
strprintf("%.777777700000000$", 1.0);
```

**Issue 2.** The following causes a stack overflow:

```
strprintf("%987654321000000:", 1);
```

**Issue 3.** The following causes a stack overflow:

```
strprintf("%1$*1$*", -11111111);
```

**Issue 4.** The following causes a `NULL` pointer dereference:

```
strprintf("%.1s", (char *)nullptr);
```

**Issue 5.** The following causes a float cast overflow:

```
strprintf("%c", -1000.0);
```

**Issue 6.** The following causes a float cast overflow followed by an invalid integer negation:

```
strprintf("%*", std::numeric_limits<double>::lowest());
```

### CODE REVIEW

**Comment by MarcoFalke** [comment]
When updating FuzzedDataProvider, please include the exact commit id that it was updated to

**Comment by practicalswift** [comment]
@MarcoFalke Good point! Done! :)

File: src/test/fuzz/strprintf.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        return;
+    }
+
+    try {
+        switch (fuzzed_data_provider.ConsumeIntegralInRange(0, 14)) {
+        case 0:
+            (void)strprintf(format_string, fuzzed_data_provider.ConsumeRandomLengthString(32));
+            break;
+        case 1:
+            (void)strprintf(format_string, fuzzed_data_provider.ConsumeRandomLengthString(32).c_str());
```
Comment (MarcoFalke) [inline src/test/fuzz/strprintf.cpp]: Why can this not be run in `case 0`? Seems odd to require the fuzzer to pick a path, when both of them can be run subsequently

File: src/test/fuzz/strprintf.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            break;
+        case 2:
+            (void)strprintf(format_string, fuzzed_data_provider.ConsumeIntegral<signed char>());
+            (void)strprintf(format_string, (signed char*)nullptr);
+            break;
+        case 3:
+            (void)strprintf(format_string, fuzzed_data_provider.ConsumeIntegral<unsigned char>());
+            (void)strprintf(format_string, (unsigned char*)nullptr);
```
Comment (MarcoFalke) [inline src/test/fuzz/strprintf.cpp]: Why are the nullptr things under `case 2` and `case 3`? Seems odd to put them there, when nothing needs to be parsed. They can be executed outside of any `case`, no?

File: src/test/fuzz/strprintf.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            break;
+        case 4:
+            (void)strprintf(format_string, fuzzed_data_provider.ConsumeIntegral<char>());
+            // Avoid triggering the following crash bug:
+            // * strprintf("%.1s", (char*)nullptr);
+            //
+            // (void)strprintf(format_string, (char*)nullptr);
```
Comment (MarcoFalke) [inline src/test/fuzz/strprintf.cpp]: Same here and for all cases below

**Review by maflcko — APPROVED** [review]
ACK

nit: in the commit that bumps the provider: Could limit the commit subject to a reasonable length, i.e. move the url to the commit body

**Comment by practicalswift** [comment]
@MarcoFalke Thanks for reviewing! Feedback addressed. Please re-review :)

**Review by chamajcpradel — APPROVED** [review]
Sirs. This is very good test.

**Comment by Empact** [comment]
Would be helpful to comment on "why" in each commit message - e.g. I was curious to look into why the `FuzzedDataProvider.h` update was necessary and found that the `ConsumeFloatingPoint` was introduced recently. https://github.com/llvm/llvm-project/commit/df3b465c9c4fed5045470d3eb5b65c14e6de71e6#diff-09bd829935aa18be3ed32f464a607c3bR211-R214

**Pull request merged**